Support of Java8 Stream as Response.Body based on a Jackson Iterator Decoder, add an option to let the decoder close the response input stream itself

phymbert opened the pull request:
This first PR commit allows to stream **only** the response body in java8 way and take benefit of Jackson token parser to decode.

I have also tried to stream request body based on InputStream as suggested by @spencergibb on [#220](https://github.com/OpenFeign/feign/issues/220 ).  But it requires additional changes on the core and maybe not appropriate implementation.
You can have a look on this [fork](https://github.com/phymbert/feign/tree/JAVA8_STREAM_INPUT).
@adriancole Please tell if it is eligible to another PR or should be included in this one or cancelled.

Fixes #514.

litianquan commented:
get it

benmanbs commented:
is there any movement on this?

adriancole commented:
@velo and I are going to go through things people note in https://github.com/OpenFeign/feign/issues/643 as important for being considered for the next release. I'm not sure if this is that or not, but mention it if you feel it is

adriancole commented:
to help guide this I suggest someone polish benchmark to latest versions of things and add jmh for this with a few response sizes

phymbert commented:
Thank you for your feedback, ,noted I will attach JMH

velo commented:
Quick question, do we really need to have one interface, one abstract class and one implementation?
Why not have `StreamDecoder` only?

benmanbs commented:
for what it's worth, I had to move forward so I ended up forking feign myself and just added `StreamDecoder`. I can open a PR with that fork if you'd like.

phymbert commented:
@velo agreed I will refactor with one class in the next 8 hours, thanks for your review. 
@benmanbs I wish this PR #587 will be merged soon with all our feedbacks, it is opened since 8 month, we can wait few days more 😁

adriancole commented on core/src/main/java/feign/codec/StreamDecoder.java:
```diff
@@ -0,0 +1,21 @@
+
+/**
+ * Custom decoder responsible to close the response stream itself.
+ * Used in java8 streaming api.
```
I think this is misleading. This has very little to do with the java 8 streams, generically speaking.  for example, normal InputStream is the same, or Iterator<Foo> or anything that does incremental parsing, right?

I don't like the type-code here as it is more about the response type than the type of the decoder. For example, the same decoder will have non streaming types dispatched to it as well.

adriancole commented on java8/src/main/java/feign/stream/Java8StreamDecoder.java:
```diff
@@ -0,0 +1,43 @@
+ * .target(GitHub.class, "https://api.github.com");
+ * </pre>
+ */
+public class Java8StreamDecoder extends AbstractStreamDecoder {
```
in most of feign code (see the HACKERS file) we do rule of three, effective java etc. One thing we can change here is get out of the eager abstraction. There's only one subtype of AbstractStreamDecoder and so we can't accurately predict the inheritance implications. IOTW, please don't expose this as it will lead to maintenance worse than a couple copy/pastes until common practice emerges. (squash the abstractStreamdecoder or hide it package protected)

adriancole commented on java8/src/main/java/feign/stream/Java8StreamDecoder.java:
```diff
@@ -0,0 +1,43 @@
+
+  private final BiFunction<Class<?>, Response, Iterator<?>> iteratorProvider;
+
+  public Java8StreamDecoder(final BiFunction<Class<?>, Response, Iterator<?>> iteratorProvider) {
```
similarly public constructors can be extremely limiting vs factory methods. Any later requirements for a Builder will have to declare different constructors. Prefer a static factory

phymbert commented:
@adriancole agreed, added option closeAfterDecode in core instead of the decoder type check. The streaming part actually expects an iterator provider to create the **java8** stream. So it is really about streaming - indeed Iterator<> based. Please note I have this fork in my project since several months and it helped a lot to handle huge collection processing by saving memory. We have no more faced OOM as before with this fix, that's the power of streaming io.InputStream http response based. Although java8 streaming based processing allows to close the stream on demand which that's finally close automatically the underlying http connection.

As promised I will add just next JVH and memory benchmark test to complete this PR.

Thank you for your double review and precious advises.

phymbert commented:
Added JMH Benchmark:
```bash
mvn clean verify -P jmh
```
Outputs:
```
# JMH version: 1.20
# VM version: JDK 1.8.0_131, VM 25.131-b11
# VM options: -server -Xms1G -Xmx1G -Dserver.url=http://localhost:54700/
# Warmup: 5 iterations, 1 s each
# Measurement: 5 iterations, 1 s each
# Timeout: 10 min per iteration
# Threads: 1 thread, will synchronize iterations
# Benchmark mode: Average time, time/op
# Benchmark: feign.stream.StreamDecoderBenchmark.feign
# Parameters: (type = list)
...
# Parameters: (type = stream)
...
# Run complete. Total time: 00:01:50
Benchmark                                                      (type)  Mode  Cnt          Score          Error   Units
StreamDecoderBenchmark.feign                                     list  avgt   15          0,677 ±        0,081    s/op
StreamDecoderBenchmark.feign:·gc.alloc.rate                      list  avgt   15        148,062 ±       12,212  MB/sec
StreamDecoderBenchmark.feign:·gc.churn.PS_Eden_Space             list  avgt   15        163,679 ±       38,391  MB/sec
StreamDecoderBenchmark.feign:·gc.churn.PS_Survivor_Space         list  avgt   15         17,037 ±       11,393  MB/sec
StreamDecoderBenchmark.feign:·gc.count                           list  avgt   15         36,000                 counts
StreamDecoderBenchmark.feign:·gc.time                            list  avgt   15       2470,000                     ms
StreamDecoderBenchmark.feign                                   stream  avgt   15          0,394 ±        0,037    s/op
StreamDecoderBenchmark.feign:·gc.alloc.rate                    stream  avgt   15        339,400 ±       27,498  MB/sec
StreamDecoderBenchmark.feign:·gc.churn.PS_Eden_Space           stream  avgt   15        348,073 ±      101,243  MB/sec
StreamDecoderBenchmark.feign:·gc.churn.PS_Survivor_Space       stream  avgt   15          0,022 ±        0,017  MB/sec
StreamDecoderBenchmark.feign:·gc.count                         stream  avgt   15         26,000                 counts
StreamDecoderBenchmark.feign:·gc.time                          stream  avgt   15         12,000                     ms
```

IMHO, clear enough stream approach is efficient for huge collection and avoid GC and Survivor space usage:

- Time : **-42%** s/op
- Survivor Space: **-99,9%** MB/s
- GC Time: **-99,5%** ms
- GC Count: **-28%** counts

What are your point of view about this benchmark ? Accurate enough ?

Can we merge 👍

adriancole commented:
on the bench, I'd probably do different sizes because people will unlikely only have huge lists

so a smaller count, like 100 as well as a larger one.

I don't think you need to show that buffering a massive list is going to be slow.

I'm more concerned about doing this via the java 8 streams api vs normal iterator.

phymbert commented:
We cannot assume what people do with feign ;-)

AFAIK total memory size of the list is linked with memory size of each listed object: but we want to decode one-by-one. You will face the same memory issue if the list is smaller but object memory size is big. By this one million cars, it is like simulating a thousand list of trucks :) is'nt ? I also think this -Xmx=1G is a real example.

About support of Iterators, agreed, I have tried to simplify the whole new API in this last commit: Java8 module is reponsible only to create the Stream. A based decoder responsible of the Iterator creation.
I have added test and example, hope it will be good enough for you @adriancole  👍

adriancole commented:
I think you misunderstand me.  Lists of size million are uncommon, but that
wasn't my main point. What I mean is that this is not a great
microbenchmark. It won't show you a realistic middle ground rather focuses
on an edge case.

You won't find inefficiencies in using this api for non "million" sized
things and also I suspect the run was slower than necessary. If you choose
to add something middle ground (typically api responses do not set limit to
1 million which is why I mention this), then you can see if there are
significant overhead differences in choosing to use this routinely for
non-million sized responses. It is a way to learn about this and if the
stream layer is adding overhead. There's a lot more code there than
iterator, and the same semantics can be used in both.

Even if you'd like not to see iterator, still you'll be able to see how
this performs in a less rediculous element count. The only reason we are
considering this is that others are interested, which is why I've asked for
you to consider mixed runs even if your apis only return one million

phymbert commented:
Thanks for clarification I appreciate your encouragement: I got your point about microbenchmark.
Updated benchmark w/ new jackson iterator decoder and ten vs million cars returned.
What do you think of this new output ?

```
# Run complete. Total time: 00:07:46

Benchmark                                                         (api)  (size)  Mode  Cnt          Score          Error   Units
StreamDecoderBenchmark.feign                                       list     few  avgt   30          0.075 ±        0.015   ms/op
StreamDecoderBenchmark.feign:·gc.alloc.rate                        list     few  avgt   30        320.863 ±       44.586  MB/sec
StreamDecoderBenchmark.feign:·gc.churn.PS_Eden_Space               list     few  avgt   30        321.503 ±       62.411  MB/sec
StreamDecoderBenchmark.feign:·gc.churn.PS_Survivor_Space           list     few  avgt   30          1.395 ±        2.862  MB/sec
StreamDecoderBenchmark.feign:·gc.count                             list     few  avgt   30         67.000                 counts
StreamDecoderBenchmark.feign:·gc.time                              list     few  avgt   30       1531.000                     ms
StreamDecoderBenchmark.feign                                       list    huge  avgt   30        707.885 ±      183.587   ms/op
StreamDecoderBenchmark.feign:·gc.alloc.rate                        list    huge  avgt   30        164.825 ±       19.498  MB/sec
StreamDecoderBenchmark.feign:·gc.churn.PS_Eden_Space               list    huge  avgt   30        166.029 ±       29.153  MB/sec
StreamDecoderBenchmark.feign:·gc.churn.PS_Old_Gen                  list    huge  avgt   30         33.097 ±       57.422  MB/sec
StreamDecoderBenchmark.feign:·gc.churn.PS_Survivor_Space           list    huge  avgt   30         40.657 ±       10.126  MB/sec
StreamDecoderBenchmark.feign:·gc.count                             list    huge  avgt   30         82.000                 counts
StreamDecoderBenchmark.feign:·gc.time                              list    huge  avgt   30       7981.000                     ms
StreamDecoderBenchmark.feign                                   iterator     few  avgt   30          0.061 ±        0.003   ms/op
StreamDecoderBenchmark.feign:·gc.alloc.rate                    iterator     few  avgt   30        458.309 ±       23.665  MB/sec
StreamDecoderBenchmark.feign:·gc.churn.PS_Eden_Space           iterator     few  avgt   30        461.582 ±       50.833  MB/sec
StreamDecoderBenchmark.feign:·gc.churn.PS_Survivor_Space       iterator     few  avgt   30          0.171 ±        0.600  MB/sec
StreamDecoderBenchmark.feign:·gc.count                         iterator     few  avgt   30         85.000                 counts
StreamDecoderBenchmark.feign:·gc.time                          iterator     few  avgt   30       1215.000                     ms
StreamDecoderBenchmark.feign                                   iterator    huge  avgt   30        522.842 ±       84.820   ms/op
StreamDecoderBenchmark.feign:·gc.alloc.rate                    iterator    huge  avgt   30        288.878 ±       36.139  MB/sec
StreamDecoderBenchmark.feign:·gc.churn.PS_Eden_Space           iterator    huge  avgt   30        289.088 ±       54.942  MB/sec
StreamDecoderBenchmark.feign:·gc.churn.PS_Survivor_Space       iterator    huge  avgt   30          0.049 ±        0.024  MB/sec
StreamDecoderBenchmark.feign:·gc.count                         iterator    huge  avgt   30         60.000                 counts
StreamDecoderBenchmark.feign:·gc.time                          iterator    huge  avgt   30        116.000                     ms
StreamDecoderBenchmark.feign                                     stream     few  avgt   30          0.058 ±        0.001   ms/op
StreamDecoderBenchmark.feign:·gc.alloc.rate                      stream     few  avgt   30        480.534 ±       10.085  MB/sec
StreamDecoderBenchmark.feign:·gc.churn.PS_Eden_Space             stream     few  avgt   30        483.290 ±       51.527  MB/sec
StreamDecoderBenchmark.feign:·gc.churn.PS_Survivor_Space         stream     few  avgt   30          0.187 ±        0.636  MB/sec
StreamDecoderBenchmark.feign:·gc.count                           stream     few  avgt   30         87.000                 counts
StreamDecoderBenchmark.feign:·gc.time                            stream     few  avgt   30       1216.000                     ms
StreamDecoderBenchmark.feign                                     stream    huge  avgt   30        524.901 ±       12.820   ms/op
StreamDecoderBenchmark.feign:·gc.alloc.rate                      stream    huge  avgt   30        266.731 ±        4.192  MB/sec
StreamDecoderBenchmark.feign:·gc.churn.PS_Eden_Space             stream    huge  avgt   30        267.846 ±       38.533  MB/sec
StreamDecoderBenchmark.feign:·gc.churn.PS_Survivor_Space         stream    huge  avgt   30          0.041 ±        0.021  MB/sec
StreamDecoderBenchmark.feign:·gc.count                           stream    huge  avgt   30         54.000                 counts
StreamDecoderBenchmark.feign:·gc.time                            stream    huge  avgt   30         90.000                     ms```

adriancole commented:
Thanks for doing this.. looking at the results, I think you can make your
point with 10 vs 100 (or worst 1000).

Then you can use a smaller unit (us) as well. When you change your code,
the smaller unit will make optimizations more obvious.

Ex in my other project custom encoder and decoder benchmarks I often use 10
vs 100 because it makes things obvious without slowing the bench or making
units weird. You can do math in your head to see if it is linear or not.
For this reason I literally say ten or hundred vs few and huge.

anyway nice work only suggestion is to change the units as seeing your
results, you don't need huge to prove your point, so let's optimize the
bench counts for what makes code change impact testing cheaper and more
obvious (smaller counts smaller units)

phymbert commented:
Based on your feedback I have introduced a new microbenchmark to compare how fast are each json arrays implementation (list, iterator, stream) for 10 vs 100 at decoders level (i.e w/o network).

Also I have moved all jmh to existing **benchmark** module and refresh it, please check if it is ok.

I would describe these two new bench this way:

1. DecoderIteratorsBenchmark, test decoders code without network. It shows up how fast is each implementation
2. FeignIteratorsBenchmark, integrated tests with network. It focus on memory consumption of different json array response processing implementations. To demonstrate memory issue with JMH GC Profiler (Cars goes to "Survivor" space), I did'nt find better way than sending huge response (again a million :) ).

I have some interrogations about results, for example why list implementation is 10 times faster in FeignIteratorsBenchmark.feign(list, 100), but DecoderIteratorsBenchmark.decode(list, 10) test shows this is the slower implementation.

How to reproduce:
```bash
mvn clean install
cd benchmark
mvn clean package
java -jar target/feign-benchmark-9.5.2-SNAPSHOT.jar
```

Results:
```
# Run complete. Total time: 00:15:54

Benchmark                                                         (api)   (size)   Mode  Cnt          Score          Error  Units
RealRequestBenchmarks.query_baseCaseUsingOkHttp                     N/A      N/A  thrpt   15      21308.756 ±      393.257  ops/s
RealRequestBenchmarks.query_feignUsingOkHttp                        N/A      N/A  thrpt   15        449.717 ±       65.886  ops/s
WhatShouldWeCacheBenchmarks.buildAndQuery_fake                      N/A      N/A  thrpt   15      20080.951 ±     1833.865  ops/s
WhatShouldWeCacheBenchmarks.buildAndQuery_fake_cachedApi            N/A      N/A  thrpt   15     607425.547 ±    19699.045  ops/s
WhatShouldWeCacheBenchmarks.buildAndQuery_fake_cachedContract       N/A      N/A  thrpt   15     115175.934 ±     1204.501  ops/s
WhatShouldWeCacheBenchmarks.buildAndQuery_fake_cachedFeign          N/A      N/A  thrpt   15      20891.088 ±      299.088  ops/s
WhatShouldWeCacheBenchmarks.parseFeignContract                      N/A      N/A  thrpt   15      24923.038 ±     3067.681  ops/s
DecoderIteratorsBenchmark.decode                                   list       10   avgt   30       6777.439 ±     1743.063  ns/op
DecoderIteratorsBenchmark.decode                                   list      100   avgt   30      38584.919 ±     5786.184  ns/op
DecoderIteratorsBenchmark.decode                               iterator       10   avgt   30       5181.077 ±      142.611  ns/op
DecoderIteratorsBenchmark.decode                               iterator      100   avgt   30      37733.995 ±     1774.236  ns/op
DecoderIteratorsBenchmark.decode                                 stream       10   avgt   30       6309.322 ±      162.222  ns/op
DecoderIteratorsBenchmark.decode                                 stream      100   avgt   30      40827.224 ±     2300.258  ns/op
FeignIteratorsBenchmark.feign                                      list      100   avgt   30     190742.456 ±    40540.121  ns/op
FeignIteratorsBenchmark.feign                                      list  1000000   avgt   30  546280895.800 ± 16307589.326  ns/op
FeignIteratorsBenchmark.feign                                  iterator      100   avgt   30    1938477.600 ±   163182.477  ns/op
FeignIteratorsBenchmark.feign                                  iterator  1000000   avgt   30  490117101.306 ± 13951727.234  ns/op
FeignIteratorsBenchmark.feign                                    stream      100   avgt   30    1997273.049 ±   169305.908  ns/op
FeignIteratorsBenchmark.feign                                    stream  1000000   avgt   30  495737640.722 ± 10265975.083  ns/op
```

With gc profiler (focus on PS_Survivor_Space and gc_time):
```
java -jar target/feign-benchmark-9.5.2-SNAPSHOT.jar -prof gc FeignIteratorsBenchmark.feign

# Run complete. Total time: 00:07:54

Benchmark                                                          (api)   (size)  Mode  Cnt          Score          Error   Units
FeignIteratorsBenchmark.feign                                       list      100  avgt   30     154605.726 ±    40294.287   ns/op
FeignIteratorsBenchmark.feign:±gc.alloc.rate                        list      100  avgt   30        276.988 ±       50.152  MB/sec
FeignIteratorsBenchmark.feign:±gc.churn.PS_Eden_Space               list      100  avgt   30        275.515 ±       82.128  MB/sec
FeignIteratorsBenchmark.feign:±gc.churn.PS_Survivor_Space           list      100  avgt   30          0.345 ±        0.666  MB/sec
FeignIteratorsBenchmark.feign:±gc.count                             list      100  avgt   30         40.000                 counts
FeignIteratorsBenchmark.feign:±gc.time                              list      100  avgt   30        189.000                     ms
FeignIteratorsBenchmark.feign                                       list  1000000  avgt   30  535357400.650 ± 12550457.145   ns/op
FeignIteratorsBenchmark.feign:±gc.alloc.rate                        list  1000000  avgt   30        260.311 ±        3.849  MB/sec
FeignIteratorsBenchmark.feign:±gc.churn.PS_Eden_Space               list  1000000  avgt   30        262.132 ±        3.873  MB/sec
FeignIteratorsBenchmark.feign:±gc.churn.PS_Survivor_Space           list  1000000  avgt   30         73.879 ±        1.101  MB/sec
FeignIteratorsBenchmark.feign:±gc.count                             list  1000000  avgt   30        120.000                 counts
FeignIteratorsBenchmark.feign:±gc.time                              list  1000000  avgt   30       3034.000                     ms
FeignIteratorsBenchmark.feign                                   iterator      100  avgt   30    2015707.425 ±   186079.768   ns/op
FeignIteratorsBenchmark.feign:±gc.alloc.rate                    iterator      100  avgt   30         26.480 ±        2.455  MB/sec
FeignIteratorsBenchmark.feign:±gc.churn.PS_Eden_Space           iterator      100  avgt   30         34.071 ±       46.304  MB/sec
FeignIteratorsBenchmark.feign:±gc.churn.PS_Survivor_Space       iterator      100  avgt   30         ? 10?±                 MB/sec
FeignIteratorsBenchmark.feign:±gc.count                         iterator      100  avgt   30          6.000                 counts
FeignIteratorsBenchmark.feign:±gc.time                          iterator      100  avgt   30        274.000                     ms
FeignIteratorsBenchmark.feign                                   iterator  1000000  avgt   30  506603976.461 ± 25947259.923   ns/op
FeignIteratorsBenchmark.feign:±gc.alloc.rate                    iterator  1000000  avgt   30        370.591 ±       25.165  MB/sec
FeignIteratorsBenchmark.feign:±gc.churn.PS_Eden_Space           iterator  1000000  avgt   30        367.887 ±       65.776  MB/sec
FeignIteratorsBenchmark.feign:±gc.churn.PS_Survivor_Space       iterator  1000000  avgt   30          0.020 ±        0.012  MB/sec
FeignIteratorsBenchmark.feign:±gc.count                         iterator  1000000  avgt   30         62.000                 counts
FeignIteratorsBenchmark.feign:±gc.time                          iterator  1000000  avgt   30        182.000                     ms
FeignIteratorsBenchmark.feign                                     stream      100  avgt   30    2066999.175 ±   233544.446   ns/op
FeignIteratorsBenchmark.feign:±gc.alloc.rate                      stream      100  avgt   30         26.009 ±        2.930  MB/sec
FeignIteratorsBenchmark.feign:±gc.churn.PS_Eden_Space             stream      100  avgt   30         28.411 ±       43.169  MB/sec
FeignIteratorsBenchmark.feign:±gc.churn.PS_Survivor_Space         stream      100  avgt   30          0.001 ±        0.002  MB/sec
FeignIteratorsBenchmark.feign:±gc.count                           stream      100  avgt   30          5.000                 counts
FeignIteratorsBenchmark.feign:±gc.time                            stream      100  avgt   30        207.000                     ms
FeignIteratorsBenchmark.feign                                     stream  1000000  avgt   30  491482226.706 ± 12741495.210   ns/op
FeignIteratorsBenchmark.feign:±gc.alloc.rate                      stream  1000000  avgt   30        384.304 ±       17.592  MB/sec
FeignIteratorsBenchmark.feign:±gc.churn.PS_Eden_Space             stream  1000000  avgt   30        384.830 ±       51.849  MB/sec
FeignIteratorsBenchmark.feign:±gc.churn.PS_Survivor_Space         stream  1000000  avgt   30          0.019 ±        0.011  MB/sec
FeignIteratorsBenchmark.feign:±gc.count                           stream  1000000  avgt   30         66.000                 counts
FeignIteratorsBenchmark.feign:±gc.time                            stream  1000000  avgt   30        190.000                     ms
```

It is the end of my vaccation, then I will not be able to be as reactive as now in the future, but of course I am welling to help. IMHO I think this PR can be merged without risk because it only add 2 new decoders  (**JacksonIteratorDecoder** + **StreamDecoder**) in existing modules (jackson + java8) and a new global option **closeAfterDecode** in core.
I hope it will be 👍. Anyway it was a nice time to learn more about feign, share with you and improve my benchmark skills. Thank you very much for that!

_lovin feign, keep it up_

adriancole commented:
Thanks for all the help. I'll closely look at all of this (and definitely change that benchmark with million in it back to 10/100 for reasons mentioned several times). If otherwise good, will merge!

adriancole commented on core/src/main/java/feign/Feign.java:
```diff
@@ -172,6 +173,21 @@ public Builder decode404() {
+     *
+     * Default is true for back compatibility.
+     *
+     * @since 9.5.2
```
will change this as we don't typically add new features on patch relase

adriancole commented on jackson/src/main/java/feign/jackson/JacksonIteratorDecoder.java:
```diff
@@ -0,0 +1,162 @@
+ *   Iterator<Contributor> contributors(@Param("owner") String owner, @Param("repo") String repo);
+ * }</code></pre>
+ *
+ * @author Pierrick HYMBERT
```
will kill author tags here and elsewhere. the only author tags we have are from code literally copied from other projects

adriancole commented on benchmark/src/main/java/feign/benchmark/FeignIteratorsBenchmark.java:
```diff
@@ -0,0 +1,167 @@
+  private HttpServer<ByteBuf, ByteBuf> server;
+
+  @Setup
+  public void setup() {
```
actually I am going to remove this benchmark. You can keep it for your own purposes, but this is a heavy test just to look at decoder performance. In hind sight "real benchmarks" in JMH aren't great as they mislead. They are at best a hint (ex use wrk or something and you'll notice a huge difference on apps assembled and impact of things together.

If we were to make a iterator based thing, I'd weave it into the main real benchmark

adriancole replied:
actually, please remove this on your own, as well the author tags. there are a fair amount of cleanups which I will explain in a summary comment.

velo commented:
> I'm not sure we should expose the Iterator option at all as it is too dangerous to use. A safe variant would close on its own on reaching the end of the stream. This one requires the caller to somehow detect this. Please change this to close the stream on its own at the end of the data.

What about having a `ClosableIterator`, make an interface that extends both `AutoCloseable` and `Iterator`.  That way modern IDE's will warn you about the need of closing the Iterator

adriancole commented:
> What about having a ClosableIterator, make an interface that extends both AutoCloseable and Iterator. That way modern IDE's will warn you about the need of closing the Iterator

auto-closeable is java 7 :) don't you love this! Also, it leaks more
feign types into apps, which might not be bad, but something to
acknowledge. Seriously I'd recommend auto-closing inside the iterator
for now, unless this is somehow impossible.

velo commented:
the risk there is someone not looping the iterator to the end and the stream being left open.

adriancole commented:
>
> the risk there is someone not looping the iterator to the end and the
> stream being left open.
>
true, though that's the same risk as a Response result, (which is larger
than the auto-buffering) correct?

On Mon, Mar 5, 2018 at 11:48 AM, Marvin Froeder <notifications@github.com>
wrote:

> the risk there is someone not looping the iterator to the end and the
> stream being left open.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/OpenFeign/feign/pull/587#issuecomment-370302373>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAD61zq9rqIdoxS7fFCsL8nyDoNX8Yzyks5tbLWRgaJpZM4O-1p_>
> .
>

benmanbs commented:
I've been largely silent on this PR because a) @phymbert requested I let him get his PR through (and I totally understand the desire to do a project like this on my own), and b) because I don't really have much of a history with Feign and figured my opinions would be better kept to myself. However, I do have one suggestion, if that's all right. 

Based on the amount of back and forth that has been going on with this PR, I think it's pretty clear that there is no consensus on the best way to handle streaming data. Maybe that's ok. Maybe that would be the library trying to answer too many questions on behalf of the users. 

My suggestion is that if you guys are trying to roll a release out shortly, why not just merge the one change with adding the `closeAfterDecode` flag, and add some comprehensive documentation about the dangers of using that options. Should someone decide that they want to use this advanced feature, they should also know that they need to close the response object when they're done with it, and they will need to write a custom decoder. Maybe a major version later can add the additional complexity this PR contains, and can be a bit more comprehensive (is JSON really the format someone is going to be using for streaming data? maybe streaming JSON?)

adriancole commented:
>
> I've been largely silent on this PR because a) @phymbert
> <https://github.com/phymbert> requested I let him get his PR through (and
> I totally understand the desire to do a project like this on my own), and
> b) because I don't really have much of a history with Feign and figured my
> opinions would be better kept to myself. However, I do have one suggestion,
> if that's all right.
>
> Based on the amount of back and forth that has been going on with this PR,
> I think it's pretty clear that there is no consensus on the best way to
> handle streaming data. Maybe that's ok. Maybe that would be the library
> trying to answer too many questions on behalf of the users.
>
> My suggestion is that if you guys are trying to roll a release out
> shortly, why not just merge the one change with adding the
> closeAfterDecode flag, and add some comprehensive documentation about the
> dangers of using that options. Should someone decide that they want to use
> this advanced feature, they should also know that they need to close the
> response object when they're done with it, and they will need to write a
> custom decoder. Maybe a major version later can add the additional
> complexity this PR contains, and can be a bit more comprehensive (is JSON
> really the format someone is going to be using for streaming data? maybe
> streaming JSON?)
>
Agreed. if only the change to core (which is literally the only thing
needed to build the extension), then it is a very small change, can be
described with the dangers of it, regardless of if someone is using it for
an Iterator, Stream, or any other lazy-parsing mechanism. Such a change
would be easy on the CHANGES file as well.

velo commented:
> though that's the same risk as a Response result

yes, but, if you are dealing with `Response` you know you need to handle things by yourself.

now, when you get a `Iterator`, you assume it's safe
thay is why I suggest a `CloseableIterator`.  People are clearly told they need to manage the resource.

benmanbs commented:
Seeing as @phymbert is the original author of this PR, I'm going to give him the opportunity to make that small PR, otherwise I'm happy to open it myself and write up some careful documentation about the dangers of using that feature.

adriancole commented:
> yes, but, if you are dealing with Response you know you need to handle things by yourself.
>
> now, when you get a Iterator, you assume it's safe
> thay is why I suggest a CloseableIterator. People are clearly told they need to manage the resource.

it is very unsafe to leave the inputstream behind an iterator dangling
or suggest not to read it until completion. no-one supports iterator
without this anyway, so it isn't required to do everything with the
type system. If you did, you know the impact, which is that feign is
now responsible for collection type which it hasn't been so far. This
means you have to be very careful not to change this api once created.

adriancole commented:
Also note that if you make this api and require it to be autocloseable
(very unnecessary), you are pinning it to JRE 1.7 only.

It is very unnecessary because Closeable will work especailly if you
are making a cover type as you can override it to hide exceptions,
assuming that's the reason for choosing AutoCloseable.

adriancole commented:
ps interesting discussion on this. I noted the rationale thing here
https://github.com/OpenFeign/feign/issues/646#issuecomment-370309800

Note that it is possible also to make a closeable iterator knowing
that which only compiles on systems that have AutoClosable handy (many
if not all even android). Just it is easier to choose a type that you
don't have to research.

At any rate, whoever merges this, should be aware and I think I'd
dumped all advice I can think of on this point :) cheers

phymbert commented:
Thank you very much for the time passed on review these changes. I have tried to push all requested updates raised by @adriancole.

I choose to go first with the option `JacksonIteratorDecoder` closes the response itself at end of the json arrays or when it fails to parse the stream. Advise in documentation this behavior.

About to separate closeAfterDecode in a dedicated PR, IMHO at this point of maturity of this PR this is too much git job. Please revert if you still think it cannot be merged all together.

adriancole commented:
We always squash anyway, so the 24 commits here can be squashed preliminarily. Then, pulling out a commit for closeAfterDecode and rebasing this overtop will not be much git work.

Food for thought.

phymbert commented:
Understood. Can it be done by commiters ? I am already far from my personal
computer, if it required actions on my side I may do it later today.

Le lun. 5 mars 2018 à 08:09, Adrian Cole <notifications@github.com> a
écrit :

> We always squash anyway, so the 24 commits here can be squashed
> preliminarily. Then, pulling out a commit for closeAfterDecode and rebasing
> this overtop will not be much git work.
>
> Food for thought.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/OpenFeign/feign/pull/587#issuecomment-370328512>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AFeaVdLv7hujSJkg78WIlWYilaUXr7yhks5tbOSRgaJpZM4O-1p_>
> .
>
-- 

Best Regards / Cordialement / С уважением,

Pierrick  *HYMBERT* / Пьеррик *ИМБЕР*

*+Pierrick <https://plus.google.com/u/0/105713262389092625238> / Skype /
LinkedIn <https://linkedin.com/pub/pierrick-hymbert/51/506/357>*

benmanbs commented:
I've opened a smaller PR with just the `closeAfterDecode` flag. I've also taken the liberty of adding a unit test for that.

adriancole commented:
Yes committers have access to rebase your branch

phymbert commented:
What happened here? Who allow to take the code of one contributor and resubmit on your own ?
Not fair at all, all these efforts to get bypassed, sad.

adriancole commented:
@phymbert I think possibly someone misunderstood you when you asked if others could help decompose the change into two commits as you were out of time.

While the same feature was achieved you may have not noticed that the flag, code and javadocs were all written independently. @benmanbs certainly did not copy/paste your code and in fact it is better as it is unit tested.

I would be less concerned about credit around the code I asked you to use to decouple this change from core. The main work you have done here is the iterator on top of it. However, these things are personal, and if stepping back you feel like you've been wronged somehow. I apologize you feel that way. If others want to make something like this, at least they can now.

OTOH, I know emotions are hot, so if you change your mind, feel free to reopen.

phymbert commented:
Noted, I rebased, squashed and will open a new PR as it has to be forced now.

adriancole commented:
no worries!

benmanbs commented:
@phymbert - I apologize if it looked like I was stepping on your toes, I was trying to decouple some of the bits of work in this PR from one another, and add some unit test coverage to the code changes. As @adriancole noted, I did not copy paste your code, I had already written my own implementation a while back (I've been maintaining a local fork with this functionality for a few weeks). I'm happy to continue help out any way I can to get your PR over the line.

phymbert commented:
I clearly got up on the wrong side of the bed! After seen your nice PR @benmanbs I do understand the move as we are all here to make a nice 9.6.0 release and build a great feign community around.
Thank you all for your time to detail your point of view.
Let's go ahead with splitted PR #649 (doNotCloseAfterDecode) and #651 (custom built-in decoders for lazy processing response).
